mathics/core/expression.py

```diff
@@ -856,16 +904,40 @@ def eval_range(indices):
             eval_range(range(len(leaves)))
             # rest_range(range(0, 0))
 
+        # Step 2: Build a new expression. Notice that leaves are given
```

Comment:
I extracted these steps and condensed them a little bit and put this [here](https://stackoverflow.com/a/8869553/546218)  in the developer's guide. 

Steps 2 and 3 are combined there because both really are about building a new expression.